docs/proposals/odo-deploy.md

```diff
@@ -0,0 +1,152 @@
+1. `odo push` - This runs the application source code using inner-loop instructions from devfile.
+1. Validate the running application is working as intended.
+1. Iterate over steps 2 and beyond (as needed). 
+1. `odo deploy` - This would build a new clean image and deploy it to the target cluster using outer-loop instructions from devfile and user-provided arguments.
+1. Validate the deployed application is working as intended.
+1. Iterate over steps 2 and beyond (as needed).
+1. Optionally, run `odo deploy delete` to clean up resources created with `odo deploy`.
+1. Push your code to Git - ready for sharing with your team and for CI/CD pipelines to take over.
+
+## User Stories
+
+### Initial build and deploy support to odo - https://github.com/openshift/odo/issues/3300
+
+## Design overview
+`odo deploy` could provide developers with a way to build a container image for their application and deploy it on a target Kubernetes deployment using the build/deploy guidance provided by the devfile.
+
+This deployment is equivalent to a development version of your production and will be using the namespace and URL information from the inner-loop. This will ensure that it is not seen as a way to deploy real workloads in production.
+
+`odo deploy delete` could provide developers with a way to clean up any existing deployment of the application. 
+
+### High-level design:
+
+#### Technical considerations:
+- The implementation would be under the experimental flag.
+- Only supported for Devfile v2.0.0 components.
+- Only supported for Kubernetes/ OpenShift targets.
+- `odo deploy` will not make any changes to the user's project.
+
+#### odo deploy 
+This command will build a container image for the application and deploy it on the target Kubernetes environment. 
+
+Flags:
+ - `--tag`: The tag to be used for the built application container image - `<registry>/<org>/<name>:<tag>` (mandatory).
+ - `service-account`: The service account for running privileged containers and push access to the image registry.
+ - `--credentials`: The credentials needed to push the image to the container image registry (mandatory).
+
+#### odo deploy delete
+This command will delete any resources created by odo deploy.
+
+### Detailed design: 
+
+### Devfile
+
+For the initial implementation, we could use devfile v2.0.0 and capture basic outer-loop information as `attributes` on the devfile. `odo` could look for  specific keys, while other tools like Che could ignore them.
+
+For example: 
+```
+attributes:
+    build-dockerfile: <URI>
+    deployment-manifest: <URI>
+```
+
+This should not need any change to the devfile v2 schema or the parser.
+
+### Dockerfile
+This could be any valid dockerfile.
+
+### Deployment Manifest
+The deployment manifest could be templated to help with replacing key bits of information:
+- PROJECT_NAME
+- CONTAINER_IMAGE
+- PORT
+
+For example: 
+```
+apiVersion: app.stacks/v1beta1
+kind: RuntimeComponent
+metadata:
+    name: PROJECT_NAME
```

Comment:
`metadata.name` generally (correct me if I'm wrong) indicates name of the resulting object that's created on k8s. In this case `RuntimeComponent`. So is the `RuntimeComponent` going to be named same as `PROJECT_NAME`? Does that mean only one `RuntimeComponent` per project?

**DISCLAIMER**: I don't know what a `RuntimeComponent` is. Asking because this looks like anomaly or maybe misinterpretation on my part.

Reply:
Apologies, I should have made this clearer. I chose an operator based deployment for the example as it shows the diff fro ma typical k8s deployment that we do during inner loop. 

https://operatorhub.io/operator/runtime-component-operator is the operator I used in this example. But, you could imagine devfile creator creating a Knative service, or other types if they wish to.

Reply:
@neeraj-laad yea, let's use a more commonly known `Kind` here please :)

Reply:
I'm not pointing to `RuntimeComponent` kind here. I'm trying to point that if we use `PROJECT_NAME` (which I assume is the name of a k8s/ocp namespace), it would probably lead to odo being able to create only one `RuntimeComponent` per project/namespace. There could be multiple applications a user is developing within the same k8s/ocp namespace, no?

Apologies if this falls into nitpick category. I felt like commenting because I understood it to be a template/default manifest that we would create when user's interested in going beyond `odo push` to `odo deploy`.

Reply:
Have updated the proposal to include multiple examples, standard Kubernetes deployment, knative service, operator based deployments. I hope this makes it clearer.